### PR TITLE
Broaden plab region search

### PR DESCRIPTION
**Pull request by kdnilsen**
Recent testing revealed that old-gen heap regions were being ignored in the search to satisfy new PLAB allocation requests.  It was discovered that many of these regions are found within ranges of the ShenandoahFreeSet that are not considered to be "is_collector_free()".

This is a first step in a two-step change.  In this patch, we broaden the search for old-gen heap regions that have available memory to include regions that are not is_collector_free.  In a second step of this improvement, we intend to restructure the implementation of the ShenandoahFreeSet to better distinguish ranges of regions that hold young-gen survivors, which ranges hold old-gen, and which ranges are intended to serve as mutator allocations.

On an Extremem workload that allocates roughly 626 M/s in a total heap size of 49G with an old-gen usage of 18.7G, we saw significant improvements compared to mainline generational Shenandoah implementation:

1. Concurrent GC passes decreased from 606 to 493 (19% improvement)
2. Degenerated GC passes decreased from 17 to 3 (82% improvement)
3. Full GCs decreased from 15 to 3 (80% improvement)
4. P50 latency for Customer Preparation Processing (CPP) improved from 1798 us to 1735 us (3.5%)
5. P100 latency for CPP improved from 25_636_285 us to 9_148_580 us (64% improvement)

Across a broad assortment of performance related CI tests, we also benefits on x86:

-74.24% extremem-phased/do_nothing_p99 p=0.00061
  Control:      2.318s  (+/-941.25ms)         80
  Test:         1.330s  (+/-  1.01s )         15

-15.70% extremem-phased/context_switch_count p=0.02032
  Control:  28188.234   (+/-5868.23  )         80
  Test:     24362.538   (+/-4260.19  )         15

-6.26% extremem-phased/do_nothing_p50 p=0.00246
  Control:    603.203us (+/- 38.32us)         80
  Test:       567.692us (+/- 50.34us)         15

And on aarch64:

+22.92% specjbb2015/sla_10000_jops p=0.01104
  Control:   2607.153   (+/-799.74  )         90
  Test:      3204.615   (+/-592.15  )         15

-5.85% extremem-phased/do_nothing_p50 p=0.00675
  Control:    608.153us (+/- 44.52us)         90
  Test:       574.538us (+/- 47.49us)         15

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Reviewers
 * [William Kemper](https://openjdk.org/census#wkemper) (@earthling-amzn - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/shenandoah pull/198/head:pull/198` \
`$ git checkout pull/198`

Update a local copy of the PR: \
`$ git checkout pull/198` \
`$ git pull https://git.openjdk.org/shenandoah pull/198/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 198`

View PR using the GUI difftool: \
`$ git pr show -t 198`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/shenandoah/pull/198.diff">https://git.openjdk.org/shenandoah/pull/198.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp [hunk 1]
```diff
@@ -242,12 +242,16 @@ bool ShenandoahConcurrentGC::collect(GCCause::Cause cause) {
       heap->set_old_evac_reserve(0);
       heap->reset_old_evac_expended();
       heap->set_promoted_reserve(0);
+      log_info(gc, ergo)("At end of Concurrent GC, old_available: " SIZE_FORMAT "%s out of total: " SIZE_FORMAT "%s,"
```
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:245]: This looks a lot like the implementation of `ShenandoahGeneration::log_status` . Could consolidate these messages and reduce logging duplicate information.

File: src/hotspot/share/gc/shenandoah/shenandoahDegeneratedGC.cpp [hunk 2]
```diff
@@ -58,9 +58,14 @@ bool ShenandoahDegenGC::collect(GCCause::Cause cause) {
   if (heap->mode()->is_generational()) {
```
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/shenandoahDegeneratedGC.cpp:58]: As above, consider using `ShenandoahGeneration::log_status`.

File: src/hotspot/share/gc/shenandoah/shenandoahFullGC.cpp [hunk 3]
```diff
@@ -176,9 +176,14 @@ void ShenandoahFullGC::op_full(GCCause::Cause cause) {
   if (heap->mode()->is_generational()) {
     size_t old_available = heap->old_generation()->available();
     size_t young_available = heap->young_generation()->available();
-    log_info(gc, ergo)("At end of Full GC, old_available: " SIZE_FORMAT "%s, young_available: " SIZE_FORMAT "%s",
+    log_info(gc, ergo)("At end of Full GC,  old_available: " SIZE_FORMAT "%s out of total: " SIZE_FORMAT "%s,"
```
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/shenandoahFullGC.cpp:179]: Consider `ShenandoahGeneration::log_status`.

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp [hunk 4]
```diff
@@ -3185,13 +3186,13 @@ void ShenandoahHeap::verify_rem_set_at_mark() {
...
               ShenandoahAsserts::print_failure(ShenandoahAsserts::_safe_all, obj, obj_addr, NULL,
-                                            "Verify init-mark remembered set violation", "object not properly registered", __FILE__, __LINE__);
+                                               "Verify init-mark remembered set violation", "object not properly registered", __FILE__, __LINE__);
             }
             obj_addr += obj->size();
           } else {
             // This object is not live so we don't verify dirty cards contained therein
-            assert(ctx->top_at_mark_start(r) == top, "Expect tams == top at start of mark.");
-            obj_addr = ctx->get_next_marked_addr(obj_addr, top);
+            assert(tams != nullptr, "If object is not live, ctx and tams should be non-null");
```
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:3194]: Might need to rebase these changes after integrating #197 .

**Review by earthling-amzn — CHANGES_REQUESTED** [review]
The workaround makes sense. Consider consolidating some of the log messages by reusing `ShenandoahGeneration::log_status`.

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp (hunk 4 above)
Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:3194]: So far, git seems to feel like there are "no conflicts".  This code is identical to what I delivered in PR197.

File: src/hotspot/share/gc/shenandoah/shenandoahFullGC.cpp (hunk 3 above)
Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/shenandoahFullGC.cpp]: I've integrated the two heap-status logging approaches.  Let me know what you think...  Thanks for suggestion.

**Review by earthling-amzn — APPROVED** [review]
Thank you for making those changes to the logging.

**Pull request closed without merging**